src/include/btree.i

```diff
@@ -1158,12 +1158,14 @@ static inline bool
 __wt_btree_truncate_active(WT_SESSION_IMPL *session, WT_REF *ref)
 {
 	WT_PAGE_DELETED *page_del;
+	uint8_t prepare_state;
 
 	if ((page_del = ref->page_del) == NULL)
 		return (false);
 	if (page_del->txnid == WT_TXN_ABORTED)
 		return (false);
-	if (page_del->prepare_state != WT_PREPARE_READY)
+	WT_ORDERED_READ(prepare_state, page_del->prepare_state);
+	if (prepare_state != WT_PREPARE_READY)
```

Comment:
1. I will merge the two functions.

2. We re-read the prepare state in `__wt_txn_upd_visible_type` after checking `__wt_txn_visible`, so a torn read of a timestamp is possible, but the result will be ignored. We will also add a new prepare state to indicate the "DONE" state instead of transitioning back to "READY".  It makes it easier to reason about `__wt_txn_upd_visible_type`.

Reply:
> We re-read the prepare state in `__wt_txn_upd_visible_type` after checking

This bothers me a little, I'm not sure if re-reading the state is our solution to timestamps where torn reads are possible, or just a side-effect of code we're writing anyway. If it's the former, maybe we need a comment?

I thought (but may be wrong), that our plan for torn timestamp reads was the `WT_WITH_TIMESTAMP_READLOCK` macro, so I'm not sure what's happening here.

Reply:
Yes, FTR the `prepare_state` design was to avoid the readlock in the (extremely common) case of reading an update.